### PR TITLE
Add authentication to the Bulk API end-point

### PR DESCRIPTION
**Pull request by jon-betts**
This is intended to only let LMS through

### CODE REVIEW

File: tests/functional/api/conftest.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import base64
```
Comment (jon-betts) [inline tests/functional/api/conftest.py]: These are heavily based off the fixtures extracted from the group tests in https://github.com/hypothesis/h/pull/5957.

This means:

 * They aren't specific to this and represent a decent "average" of the fixtures repeated throughout these API tests
 * These are probably good enough to use in the other tests too when we choose to (with appropriate modification of those tests)
 * These will be re-used when we merge: https://github.com/hypothesis/h/pull/5957

File: h/views/api/bulk.py [hunk 2]
```diff
@@ -30,13 +32,60 @@ def from_byte_stream(cls, byte_stream, executor, observer=None):
             return None
 
 
+class AuthorityCheckingExecutor(AutomaticReportExecutor):
```
Comment (jon-betts) [inline h/views/api/bulk.py]: This won't live here in the end, but it'll do for now

File: h/views/api/bulk.py [hunk 3]
```diff
@@ -30,13 +32,60 @@ def from_byte_stream(cls, byte_stream, executor, observer=None):
...
+
+        if value == self.authority:
+            return
+
+        raise InvalidDeclarationError(
+            f"The {field} '{value}' does not match the expected authority"
+        )
+
+    def _check_authority(self, data_type, body):
+        if data_type == DataType.USER:
```
Comment (jon-betts) [inline h/views/api/bulk.py]: Should these be calls on the command objects? They know where everything lives, they could be in charge of this?

File: h/auth/policy.py [hunk 4]
```diff
@@ -22,6 +22,7 @@
     ("api.users", "POST"),
     ("api.user_read", "GET"),
     ("api.user", "PATCH"),
+    ("api.bulk", "POST"),
```
Comment (seanh) [inline h/auth/policy.py]: :+1:

File: h/routes.py [hunk 5]
```diff
@@ -133,7 +133,12 @@ def includeme(config):
         traverse="/{id}",
     )
 
-    config.add_route("api.bulk", "/api/bulk", request_method="POST")
+    config.add_route(
+        "api.bulk",
+        "/api/bulk",
+        request_method="POST",
+        factory="h.traversal.BulkAPIRoot",
```
Comment (seanh) [inline h/routes.py]: :+1:

File: h/traversal/__init__.py [hunk 6]
```diff
@@ -23,6 +24,7 @@
     "Root",
     "AnnotationRoot",
     "AuthClientRoot",
+    "BulkAPIRoot",
```
Comment (seanh) [inline h/traversal/__init__.py]: :+1: We could consider removing the `__all__` from this file as I don't think it's adding anything. The `# -*- coding: utf-8 -*-` too. Might be worth separate PRs at some point to just remove all of both of those things from h

File: h/traversal/roots.py [hunk 7]
```diff
@@ -140,6 +140,13 @@ def __getitem__(self, client_id):
         return client
 
 
+class BulkAPIRoot(RootFactory):
+    """Root factory for the Bulk API."""
+
+    # Currently only LMS uses this end-point
```
Comment (seanh) [inline h/traversal/roots.py]:
```suggestion
```

I think that's explicitly clear from the ACL list itself

File: h/views/api/bulk.py (hunk 2 above)
Comment (seanh) [reply h/views/api/bulk.py]: The inheritance here makes me uncomfortable. I understand inheritance to be a tightly coupled "is a" relationship, and I don't think an `AuthorityCheckingExecutor` "is an" `AutomaticReportExecutor`.

`AutomaticReportExecutor` doesn't have any methods other than `execute_batch()` and, even if it grows more methods in future, `AuthorityCheckingExecutor` doesn't call them (doesn't call `super()` for anything other than `execute_batch()`).

It looks like what's actually going on here is that `AuthorityCheckingExecutor` is a standalone executor that makes use of an `AutomaticReportExecutor` for reporting. So I think composition might be a more appropriate pattern:

```python
class AuthorityCheckingExecutor(Executor):

    def __init__(self, authority="lms.hypothes.is"):
        self._automatic_report_executor = AutomaticReportExecutor()
        ...

    def configure(self, config):
        ...

    def execute_batch(self, command_type, data_type, default_config, batch):
        ...

        self._automatic_report_executor.execute_batch(command_type, data_type, default_config, batch)

    ...
```

I also noticed that `AutomaticReportExecutor`'s one method (`execute_batch()`) doesn't use `self`, so the class could perhaps be turned into a `create_fake_report()` function that `AuthorityCheckingExecutor` could call:

```python
class AuthorityCheckingExecutor(Executor):

    ...

    def execute_batch(self, command_type, data_type, default_config, batch):
        ...

        create_fake_report(command_type, data_type, default_config, batch)

    ...
```

File: tests/functional/api/conftest.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+
+@pytest.fixture
+def _user_for_authority(db_session, factories):
+    def _make_user(authority):
+        user = factories.User(authority=authority)
+        db_session.commit()
+        return user
+
+    return _make_user
```
Comment (seanh) [inline tests/functional/api/conftest.py]: :+1: These look useful

File: tests/functional/api/test_bulk.py [hunk 9]
```diff
@@ -7,14 +7,15 @@
 
 
 class TestBulk:
-    @pytest.mark.skip
```
Comment (seanh) [inline tests/functional/api/test_bulk.py]: FYI `xfail` might have been more appropriate here

File: h/views/api/bulk.py (hunk 2 above)
Comment (jon-betts) [reply h/views/api/bulk.py]: It won't matter in the end as the `AutomaticReportingExecutor` is only while we are faking reports, so pretty soon this won't inherit from this.

File: h/traversal/__init__.py (hunk 6 above)
Comment (jon-betts) [reply h/traversal/__init__.py]: I think they just add another thing to maintain. As an added bonus it also makes it more annoying to find the class which actually implements the code whenever you follow an import as it dumps you in a file that just imports it again.

This makes sense for a public API, but I don't think it buys much internally.

File: h/traversal/roots.py (hunk 7 above)
Comment (jon-betts) [reply h/traversal/roots.py]: I don't think that's clear to people who are less familiar with H principals and authentication.
